### PR TITLE
Implement Sleep Clock Accuracy procedure support

### PR DESCRIPTION
**Pull request by erbr-ot**
Adding support for LE REQUEST PEER SCA and peripheral handling of central update to sleep clock accuracy.

This is NOT supported in LEGACY LLCP.

### CODE REVIEW

File: tests/bluetooth/controller/common/include/helper_features.h [hunk 1]
```diff
@@ -165,7 +165,7 @@
 #define FEAT_PER_ADV_SYNC_RX 0x00
 #endif
 
-#if defined(CONFIG_MISSING)
+#if defined(CONFIG_BT_CTLR_REQUEST_PEER_SCA)
```
Comment (cvinayak) [inline tests/bluetooth/controller/common/include/helper_features.h:168]:
```suggestion
#if defined(CONFIG_BT_CTLR_SCA_UPDATE)
```

File: tests/bluetooth/controller/mock_ctrl/include/kconfig.h [hunk 2]
```diff
@@ -162,6 +162,10 @@
 #define CONFIG_BT_CTLR_DF_CTE_RX y
 #endif
 
+#ifndef CONFIG_BT_CTLR_REQUEST_PEER_SCA
+#define CONFIG_BT_CTLR_REQUEST_PEER_SCA y
+#endif
```
Comment (cvinayak) [inline tests/bluetooth/controller/mock_ctrl/include/kconfig.h:167]:
```suggestion
#ifndef CONFIG_BT_CTLR_SCA_UPDATE
#define CONFIG_BT_CTLR_SCA_UPDATE y
#endif
```

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/Kconfig [hunk 3]
```diff
@@ -172,6 +172,13 @@ config BT_PER_ADV_SYNC_TRANSFER_RECEIVER
 config BT_PER_ADV_SYNC_TRANSFER_SENDER
 	bool "Periodic Advertising Sync Transfer sender"
 	depends on !BT_CTLR || BT_CTLR_SYNC_TRANSFER_SENDER_SUPPORT
+
+config BT_REQUEST_PEER_SCA
+	bool "Sleep Clock Accuracy Req"
+	default y
+	depends on !BT_CTLR || BT_CTLR_REQUEST_PEER_SCA_SUPPORT
+	help
+	  Enable support for Bluetooth 5.1 Sleep clock accuracy Request Procedure
```
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply subsys/bluetooth/Kconfig:181]: Or use the phrasing from the 5.3 spec: "Sleep Clock Accuracy Update procedure"/"Sleep Clock Accuracy Updates"?

File: subsys/bluetooth/Kconfig [hunk 4]
```diff
@@ -172,6 +172,13 @@ config BT_PER_ADV_SYNC_TRANSFER_RECEIVER
 config BT_PER_ADV_SYNC_TRANSFER_SENDER
 	bool "Periodic Advertising Sync Transfer sender"
 	depends on !BT_CTLR || BT_CTLR_SYNC_TRANSFER_SENDER_SUPPORT
+
+config BT_REQUEST_PEER_SCA
```
Comment (rugeGerritsen) [inline subsys/bluetooth/Kconfig:176]: Do you plan adding host code for this?

**Review by rugeGerritsen — COMMENTED** [review]
What I find a bit confusing is that the spec names this feature "Sleep Clock Accuracy Updates" even though there is no requirement to actually update the SCA.

To me it would therefore be less confusing if this was named something like "Sleep Clock Accuracy exchange" or similar. Feel free to ignore this comment.

Also, I do not really understand why this feature is needed for Connected ISO other than the fact that it is mentioned as a requirement in the spec.

File: subsys/bluetooth/Kconfig (hunk 3 above)
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/Kconfig:181]: Hmm, I've 'copied' the pattern from PHY/DLE where this ..._SUPPORT is defined to refer to respective procedure LLCP support. Hence this wording here as well

File: subsys/bluetooth/controller/hci/hci.c [hunk 5]
```diff
@@ -2417,6 +2417,20 @@ static void le_reject_cis(struct net_buf *buf, struct net_buf **evt)
 #endif /* CONFIG_BT_PERIPHERAL */
 
 #if defined(CONFIG_BT_CONN)
+#if defined(CONFIG_BT_CTLR_REQUEST_PEER_SCA)
+static void le_request_peer_sca(struct net_buf *buf, struct net_buf **evt)
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/hci/hci.c:2421]: Agree - will update naming for funcs

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 6]
```diff
@@ -546,6 +546,20 @@ uint8_t ll_chm_get(uint16_t handle, uint8_t *chm)
...
+uint8_t ll_req_peer_sca(uint16_t handle)
+{
+	struct ll_conn *conn;
+
+	conn = ll_connected_get(handle);
+	if (!conn) {
+		return BT_HCI_ERR_UNKNOWN_CONN_ID;
+	}
+
+	return ull_cp_request_peer_sca(conn);
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c:559]: ack

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 7]
```diff
@@ -910,6 +921,27 @@ uint8_t ull_cp_data_length_update(struct ll_conn *conn, uint16_t max_tx_octets,
 }
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
 
+#if defined(CONFIG_BT_CTLR_REQUEST_PEER_SCA)
+uint8_t ull_cp_request_peer_sca(struct ll_conn *conn)
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c:925]: ack

File: subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h [hunk 8]
```diff
@@ -295,6 +296,12 @@ struct proc_ctx {
 			uint8_t  cis_id;
 			uint8_t error_code;
 		} cis_term;
+#if defined(CONFIG_BT_CTLR_REQUEST_PEER_SCA)
+		struct {
+			uint8_t sca;
+			uint8_t error_code;
+		} scau;
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:303]: suggest to use req_peer_sca instead

**Comment by erbr-ot** [comment]
Regarding naming a.o., as I see this. (and I hereby retract my comment above ;-))
It is a bit confusing as this procedure has two 'functions' (as described in Spec.5.3p2869). One is informing peer of update to local SCA and the other to request current SCA from peer. As 'we' do not support updating local SCA, this means that this boils down to being a 'request peer SCA' procedure.
However I agree that naming should reflect the general - so the naming re. the procedure and support (in KConfig a.o.) should use the term 'SCA_UPDATE' and not 'REQUEST_PEER_SCA'

File: subsys/bluetooth/Kconfig (hunk 4 above)
Comment (erbr-ot) [reply subsys/bluetooth/Kconfig:176]: Well, to be honest I'm a bit in doubt as to what is required for the host part? Is it 'just' to expose this to a bt_le_req_peer_sca() function?

**Comment by erbr-ot** [comment]
OK, so I've made the name change - using SCA_UPDATE all round, except for the API and complete-path where I've kept references to 'request peer sca'

File: subsys/bluetooth/Kconfig (hunk 4 above)
Comment (erbr-ot) [reply subsys/bluetooth/Kconfig]: Thinking further - I think, I'd probably leave this out of this, and then let future decide if host support is needed for this.

File: subsys/bluetooth/Kconfig [hunk 9]
```diff
@@ -172,6 +172,13 @@ config BT_PER_ADV_SYNC_TRANSFER_RECEIVER
 config BT_PER_ADV_SYNC_TRANSFER_SENDER
 	bool "Periodic Advertising Sync Transfer sender"
 	depends on !BT_CTLR || BT_CTLR_SYNC_TRANSFER_SENDER_SUPPORT
+
+config BT_SCA_UPDATE
+	bool "Sleep Clock Accuracy Update"
+	default y
```
Comment (rugeGerritsen) [inline subsys/bluetooth/Kconfig:178]: Is it intentional that BT_SCA_UPDATE is default y?

Comment (erbr-ot) [reply subsys/bluetooth/Kconfig:178]: Good question;-)
I opted to keep default y, when I created the config, but I guess it should probably be tied to ISO configs in some way. Not sure how this is done best. 
Could be 'default y if BT_ISO_CENTRAL || BT_ISO_PERIPHERAL' similar to what I did for BT_CTLR_SCA_UPDATE

Comment (erbr-ot) [reply subsys/bluetooth/Kconfig]: Done

**Comment by erbr-ot** [comment]
... and rebased

**Review by rugeGerritsen — APPROVED** [review]
Reviewed Kconfig entries

**Review by mtpr-ot — APPROVED** [review]
This is a very comprehensive change! I'm not able to verify every changed line in LLCP, but I'm confident, as all SCA update EBQ tests pass downstream. Nice work!

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 10]
```diff
@@ -1462,6 +1494,26 @@ static bool pdu_validate_cte_resp(struct pdu_data *pdu)
 }
 #endif /* CONFIG_BT_CTLR_DF_CONN_CTE_RSP */
 
+#if defined(CONFIG_BT_CTLR_SCA_UPDATE)
+static bool pdu_validate_clock_accuracy_req(struct pdu_data *pdu)
+{
+	if (pdu->len != sizeof(pdu->llctrl.sca_req) + 1) {
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:1500]:
```suggestion
	if (pdu->len != sizeof(pdu->llctrl.sca_req) + 1U) {
```

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 11]
```diff
@@ -1462,6 +1494,26 @@ static bool pdu_validate_cte_resp(struct pdu_data *pdu)
...
+		return false;
+	}
+
+	return true;
+}
+#endif /* CONFIG_BT_CTLR_DATA_LENGTH */
+
+static bool pdu_validate_clock_accuracy_rsp(struct pdu_data *pdu)
+{
+	if (pdu->len != sizeof(pdu->llctrl.sca_rsp) + 1) {
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:1510]:
```suggestion
	if (pdu->len != sizeof(pdu->llctrl.sca_rsp) + 1U) {
```

**Review by cvinayak — CHANGES_REQUESTED** [review]
minor nitpick, otherwise LGTM

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 11 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c:1510]: I suggest to accept this, as this repeats a pattern for all validation funcs.
I will follow-up with a PR changing this, using a macro instead.

**Review by cvinayak — APPROVED** [review]

**Review by thoh-ot — APPROVED** [review]

**Pull request merged**